eos-paygd/main.c

```diff
@@ -24,14 +24,62 @@
 #include <libgsystemservice/service.h>
 #include <libeos-payg/service.h>
 #include <signal.h>
+#include <linux/reboot.h>
+#include <sys/reboot.h>
 
+/* Force a reboot in situations where we are not able to enforce PAYG */
+static gboolean
+sync_and_reboot (gpointer user_data)
+{
+    g_warning ("Forcing reboot now!");
+    sync ();
+    reboot (LINUX_REBOOT_CMD_RESTART);
+    return FALSE;
+}
 
 int
 main (int   argc,
       char *argv[])
 {
   g_autoptr(GError) error = NULL;
   g_autoptr(EpgService) service = NULL;
+  guint timeout_id;
+
+  /* If eos-paygd is running from the initramfs, change the process name so
+   * that it survives the pivot to the final root filesystem. This is an
+   * (ab)use of the functionality systemd has to support storage daemons needed
+   * for mounting/unmounting the root filesystem. See
+   * https://phabricator.endlessm.com/T27037
+   */
+  if (access ("/etc/initrd-release", F_OK) >= 0)
+    argv[0][0] = '@';
+  else
+    g_debug ("eos-paygd unexpectedly running from root filesystem!");
+
+  /* Wait for the pivot to the final root so we can connect to the D-Bus
+   * daemon, but timeout after 20 minutes; otherwise we could be fooled into
+   * waiting forever. If the timeout were shorter it would make debugging the
+   * initramfs difficult. */
+  timeout_id = g_timeout_add_seconds (1200, sync_and_reboot, NULL);
+  while (access ("/etc/initrd-release", F_OK) >= 0)
+    ;
```

Comment:
I don't think a poll() call would work, because /etc/initrd-release isn't changing, systemd is just remounting filesystems so  new opens will see a different /etc which doesn't contain an initrd-release  (though if a user does touch /etc/initrd-release they have a PAYG system they can use for free 20 minutes at a time before it reboots itself - is this a concern?)

I'd love to see something polled, but I'll concede that even with a 1s sleep we're adding a max of 1 second to the boot up, and I don't know if we care about delays of that scale.

Reply:
I added a fifth of a second sleep.

You're right about the user being able to touch /etc/initrd-release. Should we go back to a shorter timeout like 10 minutes, at the risk of making debugging the initramfs difficult? The time between the initramfs starting and the root filesystem pivot should always be short right? I don't expect anyone to try to boot from a network share (and even then I think it'd be pretty short).

Reply:
If the user can touch `/etc/initrd-release`, surely they would also have enough privileges to replace the initramfs and disable PAYG entirely? i.e. They’re root.

(Forgive me; I haven’t kept up to date with the PAYG security model, so might well be completely wrong.)

Reply:
No, merely being root doesn't mean you can modify the initramfs. Or rather, you can modify it, but the Secure Boot signature check would then fail.

Reply:
Just for consideration... Other services like plymouth use additional systemd services to make the plymouth daemon aware of when this is happening.
You can add another systemd service that executes right before the switch-root, and/or you can add another systemd service to the real root that executes during early boot and tells the daemon that booting from the real root is now in progress.
Of course you'd have to define/implement a new non-dbus message passing scheme for this to be possible.